### ISSUE TITLE
[sw/silicon_creator] Bit manipulation extension for crc32

### ISSUE THREAD

**Issue by alphan** [lowRISC/opentitan#14037]
#13468 moves the current crc32 implementation to `sw/device/silicon_creator/lib`. This implementation, however, does not take advantage of the bit manipulation extension. We should 
(1) check the disassembly of the current implementation to confirm that we need a specialized bitmanip implementation, and
(2) implement a risc-v specific version that takes advantage of the bitmanip extension.

Please see the Zbr subset (`crc32.b`, `crc32c.b`, `crc32.h`, `crc32c.h`, `crc32.w`, `crc32c.w`) in the [draft spec](https://github.com/riscv/riscv-bitmanip/raw/main-history/bitmanip-0.93.pdf).

cc @arunthomas @cfrantz @moidx @luismarques

**Comment by luismarques** [comment]
Hugo had started working on some tests to check that a new bitmanip-based implementation would produce the same results as a reference plain-C implementation. He'll pick up that work again.

This should probably be implemented using inline assembly with something like [`.option arch, +zbr`](https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md#arch), to not require unratified bitmanip extensions to be enabled project-wide. That will require the new LLVM 16-based toolchain.

**Comment by dmcardle** [comment]
Here's the disassembly of crc32.o as a gist: <https://gist.github.com/dmcardle/98b1e8cb5e344f376b1dbbfe62c8f333>. I don't see any bitmanip CRC32 instructions emitted here, so let's give it a shot!

I'm planning to make a perftest so we can measure the effect on performance. (This is the technique I used when optimizing the memory functions https://github.com/lowRISC/opentitan/issues/13471.)

We'll still need some on-device tests to verify that our optimized CRC32 conforms to the pure C implementation. @HU90m: did you already start working on this part? My perftest can do a bare-minimum check, but we should probably write a handful of tests to cover unaligned pointers, short buffers, etc.

----

I generated the disassembly with this command:

```
./bazelisk.sh build --config riscv32 //sw/device/silicon_creator/lib:crc32 && /tools/riscv/bin/riscv32-unknown-elf-objdump  -D --visualize-jumps=extended-color bazel-out/k8-fastbuild/bin/sw/device/silicon_creator/lib/_objs/crc32_on_device_do_not_use_directly/crc32.o | less -R
```
